community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/PreparedQuery.scala

```diff
@@ -25,22 +25,89 @@ import org.neo4j.cypher.internal.compiler.v3_0.tracing.rewriters.RewriterConditi
 import org.neo4j.cypher.internal.frontend.v3_0.ast.{Query, Statement}
 import org.neo4j.cypher.internal.frontend.v3_0.{InputPosition, Rewriter, Scope, SemanticTable}
 
-case class PreparedQuery(statement: Statement,
-                         queryText: String,
-                         extractedParams: Map[String, Any])(val semanticTable: SemanticTable,
-                                                            val conditions: Set[RewriterCondition],
-                                                            val scopeTree: Scope,
-                                                            val notificationLogger: InternalNotificationLogger,
-                                                            val plannerName: String = "",
-                                                            val offset: Option[InputPosition] = None) {
+// A Cypher query goes through various stages of pre-processing before planning
+//
+// A prepared query captures all information that has been derived so far as part
+// of this processing
+//
+// Currently there are two types of prepared queries:
+//
+// - PreparedQuerySyntax (can be constructed without db access)
+// - PreparedQuerySemantics (construction may requires db access for resolving procedure signatures)
+//
+sealed trait PreparedQuery {
+  type SELF <: PreparedQuery
+
+  def statement: Statement
+  def queryText: String
+  def extractedParams: Map[String, Any]
+
+  def notificationLogger: InternalNotificationLogger
+  def plannerName: String
 
   def abstractQuery: AbstractQuery = statement.asQuery(notificationLogger, plannerName).setQueryText(queryText)
 
+  def rewrite(rewriter: Rewriter): SELF
```

Comment:
Just tried again... I thought so initially as well but it turns out to be incompatible with a later use in the context of calling copy